### PR TITLE
[5/N] Chunk encoding optimization: Add support of new encoding at Node

### PR DESCRIPTION
**Pull request by jianoaix**
## Why are these changes needed?
Support both gob and gnark encoding at Node.

## Checks

- [x] I've made sure the lint is passing in this PR.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, in that case, please comment that they are not relevant.
- Testing Strategy
   - [x] Unit tests
   - [ ] Integration tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: core/data.go [hunk 1]
```diff
@@ -218,10 +226,10 @@ func (b Bundle) Deserialize(data []byte) (Bundle, error) {
 	}
 	// Parse metadata
 	meta := binary.LittleEndian.Uint64(data)
-	if (meta & 0xFF) != GnarkBundleEncodingFormat {
+	if (meta >> (NumBundleHeaderBits - NumBundleEncodingFormatBits)) != GnarkBundleEncodingFormat {
 		return nil, errors.New("invalid bundle data encoding format")
 	}
-	chunkLen := meta >> 8
+	chunkLen := meta << NumBundleEncodingFormatBits >> NumBundleEncodingFormatBits
```
Comment (bxue-l2) [inline core/data.go:232]: would be good to add (), or comment on the precedence on the binary operations are from left to right

File: node/grpc/utils.go [hunk 2]
```diff
@@ -55,16 +55,25 @@ func GetBlobMessages(in *pb.StoreChunksRequest, numWorkers int) ([]*core.BlobMes
 			}
 
 			bundles := make(map[core.QuorumID]core.Bundle, len(blob.GetBundles()))
-			for j, chunks := range blob.GetBundles() {
+			for j, bundle := range blob.GetBundles() {
 				quorumID := blob.GetHeader().GetQuorumHeaders()[j].GetQuorumId()
-				bundles[uint8(quorumID)] = make([]*encoding.Frame, len(chunks.GetChunks()))
-				for k, data := range chunks.GetChunks() {
-					chunk, err := new(encoding.Frame).Deserialize(data)
+				if len(bundle.GetBundle()) > 0 {
```
Comment (bxue-l2) [inline node/grpc/utils.go:60]: can we have a get format function, it looks a bit adhoc

Comment (jianoaix) [reply node/grpc/utils.go]: There is no field to indicate the format, it's checking bundle field and then fallback to the old chunk encoding

File: node/grpc/utils.go [hunk 3]
```diff
@@ -54,17 +54,27 @@ func GetBlobMessages(in *pb.StoreChunksRequest, numWorkers int) ([]*core.BlobMes
...
-					chunk, err := new(encoding.Frame).Deserialize(data)
+				if format == core.GnarkBundleEncodingFormat {
+					bundleMsg, err := new(core.Bundle).Deserialize(bundle.GetBundle())
 					if err != nil {
 						resultChan <- err
 						return
 					}
-					bundles[uint8(quorumID)][k] = chunk
+					bundles[uint8(quorumID)] = bundleMsg
+				} else {
```
Comment (ian-shim) [inline node/grpc/utils.go:68]: nit: `else if format == core.GobBundleEncodingFormat`
then add `else` block that throws unexpected encoding format error

**Review by ian-shim — APPROVED** [review]
nice!

File: node/grpc/utils.go (hunk 3 above)
Comment (jianoaix) [reply node/grpc/utils.go]: Done.

File: node/grpc/utils.go [hunk 4]
```diff
@@ -54,17 +54,27 @@ func GetBlobMessages(in *pb.StoreChunksRequest, numWorkers int) ([]*core.BlobMes
...
+			format := node.GetBundleEncodingFormat(blob)
 			bundles := make(map[core.QuorumID]core.Bundle, len(blob.GetBundles()))
-			for j, chunks := range blob.GetBundles() {
+			for j, bundle := range blob.GetBundles() {
 				quorumID := blob.GetHeader().GetQuorumHeaders()[j].GetQuorumId()
-				bundles[uint8(quorumID)] = make([]*encoding.Frame, len(chunks.GetChunks()))
-				for k, data := range chunks.GetChunks() {
-					chunk, err := new(encoding.Frame).Deserialize(data)
+				if format == core.GnarkBundleEncodingFormat {
+					bundleMsg, err := new(core.Bundle).Deserialize(bundle.GetBundle())
```
Comment (jianoaix) [inline node/grpc/utils.go:62]: This deserialization will be 3.6x faster (https://github.com/Layr-Labs/eigenda/pull/624)

**Pull request merged**